### PR TITLE
Adds support for AMI sharing to Orgs and OUs

### LINKED ISSUE #21617: aws_ami_launch_permission: Support for OrganizationalUnitArn.N and OrganizationArn.N
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS now has support for sharing AMIs to an Organization or OrganizationUnit (OU) by specifying the ID/Arn of the Organization or OU, rather than the account number. This allows for sharing an AMI to an entire Organization or OU in its entirety, rather than continually adding account IDs as they are added to an Organization or OU.

### New or Affected Resource(s)

* aws_ami_launch_permission

### Potential Terraform Configuration

```hcl
resource "aws_ami_launch_permission" "ou_example" {
  image_id                = "ami-12345678"
  organizational_unit_arn = "arn:aws:organizations::123456789123:ou/o-foobar123/ou-1foo-2foobar"
}

resource "aws_ami_launch_permission" "org_example" {
  image_id         = "ami-12345678"
  organization_arn = "arn:aws:organizations::123456789123:root/o-foobar123/r-1foo"
}
```

### References

https://aws.amazon.com/about-aws/whats-new/2021/10/amazon-ec2-amazon-machine-images-organizations/

https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyImageAttribute.html

### PR DESCRIPTION
**Pull request by joraff**
This PR adds two new arguments to the `aws_ami_launch_permission` resource to allow for Organization and OrganizationalUnit sharing.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #21617.
Closes https://github.com/hashicorp/terraform-provider-aws/issues/11101.

Output from acceptance testing:

```shell
$ make testacc TESTARGS='-run TestAccEC2AMILaunchPermission_organization' PKG_NAME=internal/service/ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20  -run TestAccEC2AMILaunchPermission_organization -timeout 180m
=== RUN   TestAccEC2AMILaunchPermission_organization
=== PAUSE TestAccEC2AMILaunchPermission_organization
=== CONT  TestAccEC2AMILaunchPermission_organization
2022/04/06 23:24:53 Waiting for AMI ami-0dddfaa907ae8da63 to become available...
2022/04/06 23:30:23 Waiting for AMI ami-0dddfaa907ae8da63 to be deleted...
--- PASS: TestAccEC2AMILaunchPermission_organization (344.49s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ec2        346.991s
```

### CODE REVIEW

**Comment by joraff** [comment]
You will need to use an AWS account that's linked to an AWS Organization to be able to use the Organization's ARN as a data source. What should the default behavior be if the testing account is not in an Organization?

**Comment by joraff** [comment]
Regarding default behavior if the testing account is not in an org, I found the PreCheck to handle this behavior.

**Comment by mrobinson-anaplan** [comment]
Could this be added to the imagebuilder as well?  It's in the api.

**Comment by dmccaffery** [comment]
I'm not a fan of the "arn_type" property -- the underlying API is strongly-typed with UserId, Group, OrganizationalUnitArn, or OrganizationArn. I think it makes more sense to have:

```
organization_unit_arn string
organization_arn string
```

This makes the api more self-documenting and aligns with the other SDKs around this resource. It also reduces the need to lookup a case-sensitive string in the docs.

Here are the docs for the launch permission:
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_LaunchPermission.html

**Comment by joraff** [comment]
I agree with the use of distinct, strongly-types arguments as suggested above, and have implemented the changes. Tests and website docs have been added.

```sh
$ make testacc TESTARGS='-run TestAccEC2AMILaunchPermission_organization' PKG_NAME=internal/service/ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20  -run TestAccEC2AMILaunchPermission_organization -timeout 180m
=== RUN   TestAccEC2AMILaunchPermission_organization
=== PAUSE TestAccEC2AMILaunchPermission_organization
=== CONT  TestAccEC2AMILaunchPermission_organization
2022/04/06 23:24:53 Waiting for AMI ami-0dddfaa907ae8da63 to become available...
2022/04/06 23:30:23 Waiting for AMI ami-0dddfaa907ae8da63 to be deleted...
--- PASS: TestAccEC2AMILaunchPermission_organization (344.49s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ec2        346.991s
```

File: internal/service/ec2/ami_launch_permission.go [hunk 1]
```diff
@@ -112,7 +127,33 @@ func resourceAMILaunchPermissionDelete(d *schema.ResourceData, meta interface{})
...
+	idParts := strings.Split(d.Id(), ",")
+	if len(idParts) != 3 || idParts[0] == "" || idParts[1] == "" || idParts[2] == "" {
+		return nil, fmt.Errorf("Unexpected format of ID (%q), expected target,target_type,image_id", d.Id())
+	}
+
+	target_value := idParts[0]
+	target_type := idParts[1]
+	image_id := idParts[2]
+
+	if target_type == "AWS_ACCOUNT" {
```
[the remark this answers is not in the document]
Comment (joraff) [reply internal/service/ec2/ami_launch_permission.go:140]: Telling apart an account and an arn is the easy part :) Telling apart an org and an ou arn is more difficult, as they use the same arn resource type, but I suppose it is possible.

At the very least I should go back and restore the account import syntax so as not to break that feature. I switched to comma separators as that seems to be the style of most other complex imports.

[the remark this answers is not in the document]
Comment (joraff) [reply internal/service/ec2/ami_launch_permission.go:140]: Scratch that, I had not really looked at the arns closely and they are easily distinguishable after all.

**Comment by joraff** [comment]
Import syntax has been reverted. Needed to happen to not break compatibility with the way it is presently, `account-id/image-id`. However, both a forward slash and a comma now work for account imports, and a command is the way forward for all resource types - slashes don't mix well with arns.

The resource type is now inferred based on a regex of digits for an account number, or from parsing the arn.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTS=TestAccEC2AMILaunchPermission_ PKG=ec2 ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 2 -run='TestAccEC2AMILaunchPermission_'  -timeout 180m
=== RUN   TestAccEC2AMILaunchPermission_basic
=== PAUSE TestAccEC2AMILaunchPermission_basic
=== RUN   TestAccEC2AMILaunchPermission_disappears
=== PAUSE TestAccEC2AMILaunchPermission_disappears
=== RUN   TestAccEC2AMILaunchPermission_Disappears_ami
=== PAUSE TestAccEC2AMILaunchPermission_Disappears_ami
=== RUN   TestAccEC2AMILaunchPermission_group
=== PAUSE TestAccEC2AMILaunchPermission_group
=== RUN   TestAccEC2AMILaunchPermission_organizationARN
=== PAUSE TestAccEC2AMILaunchPermission_organizationARN
=== RUN   TestAccEC2AMILaunchPermission_organizationalUnitARN
=== PAUSE TestAccEC2AMILaunchPermission_organizationalUnitARN
=== CONT  TestAccEC2AMILaunchPermission_basic
=== CONT  TestAccEC2AMILaunchPermission_group
--- PASS: TestAccEC2AMILaunchPermission_group (361.35s)
=== CONT  TestAccEC2AMILaunchPermission_Disappears_ami
--- PASS: TestAccEC2AMILaunchPermission_basic (361.98s)
=== CONT  TestAccEC2AMILaunchPermission_organizationalUnitARN
    acctest.go:688: skipping tests; this AWS account must not be an existing member of an AWS Organization
--- SKIP: TestAccEC2AMILaunchPermission_organizationalUnitARN (0.23s)
=== CONT  TestAccEC2AMILaunchPermission_organizationARN
--- PASS: TestAccEC2AMILaunchPermission_Disappears_ami (350.94s)
=== CONT  TestAccEC2AMILaunchPermission_disappears
--- PASS: TestAccEC2AMILaunchPermission_organizationARN (355.95s)
--- PASS: TestAccEC2AMILaunchPermission_disappears (355.58s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	1071.663s
```

```console
% make testacc TESTS=TestAccEC2AMILaunchPermission_organizationalUnitARN PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccEC2AMILaunchPermission_organizationalUnitARN'  -timeout 180m
=== RUN   TestAccEC2AMILaunchPermission_organizationalUnitARN
=== PAUSE TestAccEC2AMILaunchPermission_organizationalUnitARN
=== CONT  TestAccEC2AMILaunchPermission_organizationalUnitARN
--- PASS: TestAccEC2AMILaunchPermission_organizationalUnitARN (345.34s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	349.518s
```

**Comment by ewbankkit** [comment]
@joraff @grahamc Thanks for the contribution :tada: :clap:.

**Pull request merged**